### PR TITLE
Corrected import of static and anonymous functions.

### PR DESCRIPTION
**Pull request by balazske**

### CODE REVIEW

File: unittests/AST/ASTImporterTest.cpp [hunk 1]
```diff
@@ -55,22 +57,31 @@ const StringRef DeclToVerifyID = "declToVerify";
 // Common base for the different families of ASTImporter tests that are
 // parameterized on the compiler options which may result a different AST. E.g.
 // -fms-compatibility or -fdelayed-template-parsing.
-struct ParameterizedTestsFixture : ::testing::TestWithParam<ArgVector> {
+class LanguageSpecificTest : public ::testing::Test {
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: This seems like a better name we had previously, I also like the abstraction on `getExtraArgs`.

Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: But it is not really language specific but rather compiler option specific. So perhaps `CompilerOptionSpecificTest` is more descriptive.

File: unittests/AST/ASTImporterTest.cpp [hunk 2]
```diff
@@ -461,6 +479,13 @@ class ASTImporterTestBase : public ParameterizedTestsFixture {
   }
 };
 
+class ASTImporterLangParamTestBase
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: Perhaps this should be part of `ASTImporterTestBase`.  I don't see too much benefits for having a separate class for this. `TestImportBase` is also derived from `::testing::WithParamInterface<ArgVector>`, to be consistent, I think `ASTImporterTestBase` should derive from that too.

File: unittests/AST/ASTImporterTest.cpp [hunk 3]
```diff
@@ -11,6 +11,8 @@
 //
 //===----------------------------------------------------------------------===//
 
+#define GTEST_HAS_COMBINE 1
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: Why is this needed? Because of the multiple inheritance of the test fixtures? Could you please add a comment in the code?

Comment (balazske) [reply unittests/AST/ASTImporterTest.cpp]: I do not like this `#define` but could not find other fast solution to make it work (without this define `Combine` is not avaliable).

File: lib/AST/ASTImporter.cpp [hunk 4]
```diff
@@ -3001,32 +3001,46 @@ ExpectedDecl ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
...
-            FoundByLookup = FoundFunction;
-            break;
-          }
 
-          // FIXME: Check for overloading more carefully, e.g., by boosting
-          // Sema::IsOverload out to the AST library.
+        bool VisibilityMatch = true;
+        if (D->hasExternalFormalLinkage())
+          VisibilityMatch = FoundFunction->hasExternalFormalLinkage();
+        else {
```
Comment (martong) [inline lib/AST/ASTImporter.cpp]: I think you could simplify this by uplifting the TU match and by starting with the anonymous namespace case first, something like:
```
        else if (Importer.GetFromTU(FoundFunction) == D->getTranslationUnitDecl()) {
          if (D->isInAnonymousNamespace())
            VisibilityMatch = FoundFunction->isInAnonymousNamespace()                  
          else
             VisibilityMatch = !FoundFunction->isInAnonymousNamespace() &&
                              !FoundFunction->hasExternalFormalLinkage()            
                           

```

File: unittests/AST/ASTImporterTest.cpp (hunk 2 above)
Comment (balazske) [reply unittests/AST/ASTImporterTest.cpp]: We can not add the test parameter type here if it should be possible to have tests with different kinds of parameters (and the same base class for the fixture). The test can have only one parameter type and no derived test can extend the supported parameter types.

File: lib/AST/ASTImporter.cpp [hunk 5]
```diff
@@ -3001,32 +3001,46 @@ ExpectedDecl ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
...
-                FoundFunction->hasBody(Definition))
-              return Importer.MapImported(
-                  D, const_cast<FunctionDecl *>(Definition));
-            FoundByLookup = FoundFunction;
-            break;
-          }
 
-          // FIXME: Check for overloading more carefully, e.g., by boosting
-          // Sema::IsOverload out to the AST library.
+        bool VisibilityMatch = true;
```
Comment (martong) [inline lib/AST/ASTImporter.cpp]: Could we simplify line 3005-3018 this way?
```
bool VisibilityMatch = D->hasExternalFormalLinkage() == FoundFunction->hasExternalFormalLinkage() && 
  D->isInAnonymousNamespace() == FoundFunction->isInAnonymousNamespace() &&
  Importer.GetFromTU(FoundFunction) == D->getTranslationUnitDecl();
```

Comment (martong) [reply lib/AST/ASTImporter.cpp]: Also perhaps the whole could go in a separate function to make smaller the already long VisitFunctionDecl.

File: unittests/AST/ASTImporterTest.cpp [hunk 6]
```diff
@@ -2605,6 +2626,112 @@ TEST_P(ImportFunctions,
...
+          std::tuple<ArgVector, const char *>> {
+protected:
+  ArgVector getExtraArgs() const override { return std::get<0>(GetParam()); }
+  std::string getCode() const { return std::get<1>(GetParam()); }
+};
+
+class ImportFunctionsVisibility
+    : public ASTImporterTestBase,
+      public ::testing::WithParamInterface<
+          std::tuple<ArgVector, std::tuple<const char *, const char *, bool>>> {
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: We use a tuple in a tuple because otherwise combine would not work, right?

File: unittests/AST/ASTImporterTest.cpp [hunk 7]
```diff
@@ -2605,6 +2626,112 @@ TEST_P(ImportFunctions,
...
+const char *ExternF = "void f();";
+const char *StaticF = "static void f();";
+const char *AnonF = "namespace { void f(); }";
+
+TEST_P(ImportFunctionsVisibilityChain, ImportChain) {
+  std::string Code = getCode() + getCode();
+  TranslationUnitDecl *FromTu = getTuDecl(Code, Lang_CXX, "input0.c");
+
+  auto *FromF0 = FirstDeclMatcher<FunctionDecl>().match(
+      FromTu, functionDecl(hasName("f")));
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: You could have a `Pattern` variable for the matcher so you could use that later too.

File: unittests/AST/ASTImporterTest.cpp [hunk 8]
```diff
@@ -2605,6 +2626,112 @@ TEST_P(ImportFunctions,
...
+      public ::testing::WithParamInterface<
+          std::tuple<ArgVector, std::tuple<const char *, const char *, bool>>> {
+protected:
+  ArgVector getExtraArgs() const override { return std::get<0>(GetParam()); }
+  std::string getCode0() const { return std::get<0>(std::get<1>(GetParam())); }
+  std::string getCode1() const { return std::get<1>(std::get<1>(GetParam())); }
+  bool shouldBeLinked() const { return std::get<2>(std::get<1>(GetParam())); }
+};
+
+const char *ExternF = "void f();";
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: auto * ?

File: unittests/AST/ASTImporterTest.cpp [hunk 9]
```diff
@@ -2605,6 +2626,112 @@ TEST_P(ImportFunctions,
...
+INSTANTIATE_TEST_CASE_P(ParameterizedTests, ImportFunctionsVisibilityChain,
+                        ::testing::Combine(DefaultTestValuesForRunOptions,
+                                           ImportChainParams), );
+
+TEST_P(ImportFunctionsVisibility, ImportAfter) {
+  TranslationUnitDecl *ToTu = getToTuDecl(getCode0(), Lang_CXX);
+  TranslationUnitDecl *FromTu = getTuDecl(getCode1(), Lang_CXX, "input1.cc");
+
+  auto *ToF0 =
+      FirstDeclMatcher<FunctionDecl>().match(ToTu, functionDecl(hasName("f")));
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: Perhaps the `Pattern` definition could go into the test Fixture itself (ImportFunctionsVisibility).

File: unittests/AST/ASTImporterTest.cpp [hunk 10]
```diff
@@ -2605,6 +2626,112 @@ TEST_P(ImportFunctions,
...
+  EXPECT_NE(ToF0, ToF1);
+
+  if (shouldBeLinked())
+    EXPECT_EQ(ToF1->getPreviousDecl(), ToF0);
+  else
+    EXPECT_FALSE(ToF1->getPreviousDecl());
+}
+
+auto ImportAfterParams =
+    ::testing::Values(std::make_tuple(ExternF, ExternF, true),
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: `/*shouldBeLinked=*/false` would make it easier to read.

File: lib/AST/ASTImporter.cpp (hunk 5 above)
Comment (balazske) [reply lib/AST/ASTImporter.cpp]: If we experiment with formal logic expressions probably we can find a more simple way to use instead of `if` but the current code may be more easy to understand. Testing for "everything is equal" (like above) is not correct.

File: unittests/AST/ASTImporterTest.cpp (hunk 6 above)
Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: I understand now that the tuple in the tuple is needed for the test combination,  but this deserves a descriptive comment I think.

**Review by martong — COMMENTED** [review]
Nice work!

File: unittests/AST/ASTImporterTest.cpp (hunk 2 above)
Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: OK.

**Comment by martong** [comment]
Could you please check if the test given in https://github.com/Ericsson/clang/issues/335 passes with the modifications?

**Comment by martong** [comment]
Looks good, I think you can proceed and push this onto your main branch.

**Comment by balazske** [comment]
Last comments are resolved, added the test in #335 (does pass).

File: unittests/AST/ASTImporterTest.cpp [hunk 11]
```diff
@@ -2644,20 +2654,25 @@ class ImportFunctionsVisibility
   std::string getCode0() const { return std::get<0>(std::get<1>(GetParam())); }
   std::string getCode1() const { return std::get<1>(std::get<1>(GetParam())); }
   bool shouldBeLinked() const { return std::get<2>(std::get<1>(GetParam())); }
+
+  static decltype(functionDecl()) PatternF;
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: Can't you have a simple non-static data member? Than it wouldn't be necessary to initialize the variable out-of-class.
```
auto PatternF = functionDecl(hasName("f")); // C++14
decltype(functionDecl()) PatternF = functionDecl(hasName("f")); // C++11
```

Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]:
```
auto PatternF = functionDecl(hasName("f"));
```
I think I am wrong, `auto` may not work even in C++14.

File: unittests/AST/ASTImporterTest.cpp [hunk 12]
```diff
@@ -2717,27 +2728,33 @@ TEST_P(ImportFunctionsVisibility, ImportAfterImport) {
     EXPECT_FALSE(ToF1->getPreviousDecl());
 }
 
+bool ExpectLink = true;
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: const?

This is just a very minor comment from this point: perhaps an `enum class` is even better for this purpose than `bool`. Something like `enum class Linkage { Linkable, NotLinkable }; `

File: unittests/AST/ASTImporterTest.cpp (hunk 11 above)
Comment (balazske) [reply unittests/AST/ASTImporterTest.cpp]: I got error "field initializer is not constant", so this may work only with constructor.

File: unittests/AST/ASTImporterTest.cpp (hunk 12 above)
Comment (balazske) [reply unittests/AST/ASTImporterTest.cpp]: I wanted not more "complications" with this, the enum should be inside the fixture class and names of the members would be too long in the initializer list, and in this very simple case it is not necessary. If a new (third) value is used it can be useful.

**Comment by balazske** [comment]
This looks not applicable any more, probably the code is added in some way already. (But we can not see now why this was not closed previously.)
